File: python/2015/day08a/day08a.py
#!/usr/bin/env python3

import sys

def usage():
    progname = sys.argv[0]
    print(f'usage: {progname} <input file>')
    sys.exit(1)

def process(filename):
    result = 0
    with open(filename, 'r') as infile:
        for line in infile:
            code_len = len(line)
            mem_len = 0
            i = 1
            while i < len(line) - 1:
                if line[i] == '\\':
                    if line[i + 1] == '\\' or line[i + 1] == '"':
                        i += 2
                    elif line[i + 1] == 'x':
                        i += 4
                    else:
                        i += 1
                else:
                    i += 1
                mem_len += 1
            result += code_len - mem_len
    return result

def main():
    if len(sys.argv) < 2:
        usage()
    filename = sys.argv[1]
    result = process(filename)
    print(f'result = {result}')

if __name__ == '__main__':
    main()

File: perl/2015/day08a/day08a.pl
#!/usr/bin/env perl

use strict;
use warnings;

sub usage {
    my $progname = $0;
    print "usage: $progname <input file>\n";
    exit 1;
}

sub process {
    my ($filename) = @_;
    my $result = 0;

    open ( my $infile, '<', $filename ) or die "cannot open file: $filename: $!";
    while ( my $line = <$infile> ) {
        my $codeLen = length($line);
        my $memLen = 0;
        my $i = 1;
        while ( $i < length($line) - 1 ) {
            if ( substr($line, $i, 1) eq '\\' ) {
                if ( substr($line, $i + 1, 1) eq '\\' || substr($line, $i + 1, 1) eq '"' ) {
                    $i += 2;
                } elsif ( substr($line, $i + 1, 1) eq 'x' ) {
                    $i += 4;
                } else {
                    $i += 1;
                }
            } else {
                $i += 1;
            }
            $memLen += 1;
        }
        $result += $codeLen - $memLen;
    }
    close($infile);

    return $result;
}

sub main {
    if ( @ARGV < 1 ) {
        usage();
    }
    my $filename = $ARGV[0];
    my $result = process($filename);
    print "result = $result\n";
}

main();
